File: docs/libre-et-accomplis/scaffolding/nodejs/agentNutrition.mjs
/**
 * Libre & Accomplis — Agent Nutrition (scaffolding Node.js).
 *
 * Règles de sécurité validées par l'Expert en Santé :
 *   - Jamais de recette contenant un allergène de l'utilisateur.
 *   - En l'absence de recette compatible, répondre success=false plutôt
 *     que de proposer une recette à risque.
 */
import { TAGS_PAR_OBJECTIF } from './baseConnaissances.mjs';

export class AgentNutrition {
  constructor(base) {
    this.base = base;
  }

  proposerRecette(userId) {
    const utilisateur = this.base.getUtilisateur(userId);
    const recettes = this.base.getRecettes({ allergies: utilisateur.allergies ?? [] });
    const recette = this.#selectionnerMeilleureRecette(recettes, utilisateur.objectifs ?? []);

    if (!recette) {
      return { success: false, raison: 'aucune_recette_compatible' };
    }
    return {
      id: recette.id,
      nom: recette.nom,
      ingredients: recette.ingredients,
      tags: recette.tags,
      success: true,
    };
  }

  #selectionnerMeilleureRecette(recettes, objectifs) {
    if (recettes.length === 0) return null;

    const tagsRecherches = new Set(objectifs.flatMap((o) => TAGS_PAR_OBJECTIF[o] ?? []));
    const score = (r) => r.tags.filter((t) => tagsRecherches.has(t)).length;
    return recettes.reduce((best, r) => (score(r) > score(best) ? r : best));
  }
}

File: docs/libre-et-accomplis/scaffolding/python/agent_nutrition.py
"""Libre & Accomplis — Agent Nutrition (scaffolding).

Propose des recettes adaptées au profil de l'utilisateur (allergies,
objectifs). Règles de sécurité validées par l'Expert en Santé :
  - Jamais de recette contenant un allergène de l'utilisateur.
  - En l'absence de recette compatible, répondre success=False plutôt
    que de proposer une recette à risque.
"""
from base_connaissances import BaseConnaissances, TAGS_PAR_OBJECTIF


class AgentNutrition:
    def __init__(self, base: BaseConnaissances):
        self.base = base

    def proposer_recette(self, user_id: int) -> dict:
        """Propose la recette la plus adaptée à l'utilisateur."""
        utilisateur = self.base.get_utilisateur(user_id)
        recettes = self.base.get_recettes(
            allergies=utilisateur.get("allergies", []),
            objectifs=utilisateur.get("objectifs", []),
        )
        recette = self._selectionner_meilleure_recette(
            recettes, utilisateur.get("objectifs", [])
        )
        if recette is None:
            return {"success": False, "raison": "aucune_recette_compatible"}

        return {
            "id": recette["id"],
            "nom": recette["nom"],
            "ingredients": recette["ingredients"],
            "tags": recette["tags"],
            "success": True,
        }

    def _selectionner_meilleure_recette(self, recettes: list, objectifs: list):
        """Sélectionne la recette dont les tags correspondent le mieux aux objectifs."""
        if not recettes:
            return None

        tags_recherches = set()
        for objectif in objectifs:
            tags_recherches.update(TAGS_PAR_OBJECTIF.get(objectif, []))

        def score(recette):
            return len(tags_recherches.intersection(recette["tags"]))

        return max(recettes, key=score)
